File: packages/columine/src/vm/undo_log.zig
// =============================================================================
// Undo Log — Typed rollback helpers via FlatHashTable
// =============================================================================
//
// Provides typed rollback functions that use FlatHashTable instead of raw
// pointer arithmetic. Called by rollbackEntry in vm.zig.
//
// Bitmap and TTL rollback remain in vm.zig (depend on roaring + eviction index).

const std = @import("std");
const types = @import("types.zig");
const ht = @import("hash_table.zig");

const TOMBSTONE = types.TOMBSTONE;
const SlotMeta = types.SlotMeta;
const HashMap = ht.FlatHashTable(u32);
const HashSet = ht.HashSet;

/// Timestamp side-array (after keys + values in SoA layout).
fn getTimestamps(state_base: [*]u8, meta: SlotMeta) [*]f64 {
    return @ptrCast(@alignCast(state_base + meta.offset + meta.capacity * 8));
}

// =============================================================================
// HashMap rollback
// =============================================================================

/// Rollback MAP_INSERT: tombstone key, decrement size.
pub fn rollbackMapInsert(state_base: [*]u8, meta: SlotMeta, key: u32) bool {
    const tbl = HashMap.bindSlot(state_base, meta);
    const pos = tbl.find(key) orelse return false;
    tbl.keys[pos] = TOMBSTONE;
    tbl.size_ptr.* -= 1;
    return true;
}

/// Rollback MAP_UPDATE: restore previous value + optional timestamp.
pub fn rollbackMapUpdate(state_base: [*]u8, meta: SlotMeta, key: u32, prev_value: u32, prev_ts_bits: u64) bool {
    const tbl = HashMap.bindSlot(state_base, meta);
    const pos = tbl.find(key) orelse return false;
    tbl.entries[pos] = prev_value;
    if (meta.hasHashMapTimestampStorage()) {
        getTimestamps(state_base, meta)[pos] = @bitCast(prev_ts_bits);
    }
    return true;
}

/// Rollback MAP_DELETE: restore key + value + optional timestamp, increment size.
pub fn rollbackMapDelete(state_base: [*]u8, meta: SlotMeta, key: u32, prev_value: u32, prev_ts_bits: u64) bool {
    const tbl = HashMap.bindSlot(state_base, meta);
    const probe = tbl.findInsert(key) orelse return false;
    if (probe.found) return false;
    tbl.keys[probe.pos] = key;
    tbl.entries[probe.pos] = prev_value;
    if (meta.hasHashMapTimestampStorage()) {
        getTimestamps(state_base, meta)[probe.pos] = @bitCast(prev_ts_bits);
    }
    tbl.size_ptr.* += 1;
    return true;
}

// =============================================================================
// Aggregate rollback
// =============================================================================

/// Rollback AGG_UPDATE: restore previous value (f64/i64) and count.
/// Layout: [value: u64 @ offset] [count: u64 @ offset+8] — 16-byte SUM/MIN/MAX/AVG slot.
/// prev_value_bits holds the previous count (truncated to u32, extended back to u64).
/// prev_aux holds the previous value bits (f64/i64 bitcast to u64).
pub fn rollbackAggUpdate(state_base: [*]u8, meta: SlotMeta, prev_count: u32, prev_val_bits: u64) void {
    const val_ptr: *u64 = @ptrCast(@alignCast(state_base + meta.offset));
    const count_ptr: *u64 = @ptrCast(@alignCast(state_base + meta.offset + 8));
    val_ptr.* = prev_val_bits;
    count_ptr.* = prev_count;
}

/// Rollback COUNT_UPDATE: restore previous count.
/// Layout: [count: u64 @ offset] — 8-byte COUNT-only slot.
pub fn rollbackCountUpdate(state_base: [*]u8, meta: SlotMeta, prev_count: u32) void {
    const count_ptr: *u64 = @ptrCast(@alignCast(state_base + meta.offset));
    count_ptr.* = prev_count;
}

// =============================================================================
// HashSet rollback
// =============================================================================

/// Rollback SET_INSERT: tombstone element, decrement size.
pub fn rollbackSetInsert(state_base: [*]u8, meta: SlotMeta, elem: u32) bool {
    const tbl = HashSet.bindSlot(state_base, meta);
    const pos = tbl.find(elem) orelse return false;
    tbl.keys[pos] = TOMBSTONE;
    tbl.size_ptr.* -= 1;
    return true;
}

/// Rollback SET_DELETE: restore element, increment size.
pub fn rollbackSetDelete(state_base: [*]u8, meta: SlotMeta, elem: u32) bool {
    const tbl = HashSet.bindSlot(state_base, meta);
    const probe = tbl.findInsert(elem) orelse return false;
    if (probe.found) return false;
    tbl.keys[probe.pos] = elem;
    tbl.size_ptr.* += 1;
    return true;
}

// =============================================================================
// Tests
// =============================================================================

const testing = std.testing;
const EMPTY_KEY = types.EMPTY_KEY;

test "rollbackMapInsert — tombstones key and decrements size" {
    var state: [4096]u8 align(8) = [_]u8{0} ** 4096;
    const cap: u32 = 16;
    const slot_offset: u32 = types.STATE_HEADER_SIZE + types.SLOT_META_SIZE;
    const meta_base = types.STATE_HEADER_SIZE;

    std.mem.writeInt(u32, state[meta_base..][0..4], slot_offset, .little);
    std.mem.writeInt(u32, state[meta_base + 4 ..][0..4], cap, .little);
    std.mem.writeInt(u32, state[meta_base + 8 ..][0..4], 0, .little);
    state[meta_base + 12] = 0x40; // HASHMAP, no_timestamps=true

    const keys: [*]u32 = @ptrCast(@alignCast(&state[slot_offset]));
    for (0..cap) |j| keys[j] = EMPTY_KEY;

    const meta = types.getSlotMeta(&state, 0);
    const tbl = HashMap.bindSlot(&state, meta);

    _ = tbl.upsert(42, 100);
    try testing.expectEqual(@as(u32, 1), tbl.size());

    try testing.expect(rollbackMapInsert(&state, meta, 42));
    try testing.expectEqual(@as(u32, 0), tbl.size());
    try testing.expect(!tbl.contains(42));
}

test "rollbackMapUpdate — restores previous value" {
    var state: [4096]u8 align(8) = [_]u8{0} ** 4096;
    const cap: u32 = 16;
    const slot_offset: u32 = types.STATE_HEADER_SIZE + types.SLOT_META_SIZE;
    const meta_base = types.STATE_HEADER_SIZE;

    std.mem.writeInt(u32, state[meta_base..][0..4], slot_offset, .little);
    std.mem.writeInt(u32, state[meta_base + 4 ..][0..4], cap, .little);
    std.mem.writeInt(u32, state[meta_base + 8 ..][0..4], 0, .little);
    state[meta_base + 12] = 0x40;

    const keys: [*]u32 = @ptrCast(@alignCast(&state[slot_offset]));
    for (0..cap) |j| keys[j] = EMPTY_KEY;

    const meta = types.getSlotMeta(&state, 0);
    const tbl = HashMap.bindSlot(&state, meta);

    _ = tbl.upsert(42, 100);
    _ = tbl.upsert(42, 200);
    try testing.expectEqual(@as(u32, 200), tbl.get(42).?.*);

    try testing.expect(rollbackMapUpdate(&state, meta, 42, 100, 0));
    try testing.expectEqual(@as(u32, 100), tbl.get(42).?.*);
}

test "rollbackMapDelete — restores key+value and increments size" {
    var state: [4096]u8 align(8) = [_]u8{0} ** 4096;
    const cap: u32 = 16;
    const slot_offset: u32 = types.STATE_HEADER_SIZE + types.SLOT_META_SIZE;
    const meta_base = types.STATE_HEADER_SIZE;

    std.mem.writeInt(u32, state[meta_base..][0..4], slot_offset, .little);
    std.mem.writeInt(u32, state[meta_base + 4 ..][0..4], cap, .little);
    std.mem.writeInt(u32, state[meta_base + 8 ..][0..4], 0, .little);
    state[meta_base + 12] = 0x40;

    const keys: [*]u32 = @ptrCast(@alignCast(&state[slot_offset]));
    for (0..cap) |j| keys[j] = EMPTY_KEY;

    const meta = types.getSlotMeta(&state, 0);
    const tbl = HashMap.bindSlot(&state, meta);

    // Insert then manually tombstone (simulating what the VM does on delete)
    _ = tbl.upsert(42, 100);
    const pos = tbl.find(42).?;
    tbl.keys[pos] = TOMBSTONE;
    tbl.size_ptr.* -= 1;
    try testing.expect(!tbl.contains(42));

    // Rollback the delete — should restore the key+value
    try testing.expect(rollbackMapDelete(&state, meta, 42, 100, 0));
    try testing.expect(tbl.contains(42));
    try testing.expectEqual(@as(u32, 100), tbl.get(42).?.*);
    try testing.expectEqual(@as(u32, 1), tbl.size());
}

test "rollbackSetInsert — tombstones element" {
    var state: [4096]u8 align(8) = [_]u8{0} ** 4096;
    const cap: u32 = 16;
    const slot_offset: u32 = types.STATE_HEADER_SIZE + types.SLOT_META_SIZE;
    const meta_base = types.STATE_HEADER_SIZE;

    std.mem.writeInt(u32, state[meta_base..][0..4], slot_offset, .little);
    std.mem.writeInt(u32, state[meta_base + 4 ..][0..4], cap, .little);
    std.mem.writeInt(u32, state[meta_base + 8 ..][0..4], 0, .little);
    state[meta_base + 12] = @intFromEnum(types.SlotType.HASHSET);

    const keys: [*]u32 = @ptrCast(@alignCast(&state[slot_offset]));
    for (0..cap) |j| keys[j] = EMPTY_KEY;

    const meta = types.getSlotMeta(&state, 0);
    const tbl = HashSet.bindSlot(&state, meta);

    _ = tbl.insertKey(42);
    try testing.expect(tbl.contains(42));

    try testing.expect(rollbackSetInsert(&state, meta, 42));
    try testing.expect(!tbl.contains(42));
    try testing.expectEqual(@as(u32, 0), tbl.size());
}

test "rollback — AGG_UPDATE restores previous f64 value and count" {
    var state: [4096]u8 align(8) = [_]u8{0} ** 4096;
    const meta_base = types.STATE_HEADER_SIZE;
    // AGGREGATE SUM slot: 16 bytes (8 value + 8 count), offset just past metadata
    const slot_offset: u32 = types.STATE_HEADER_SIZE + types.SLOT_META_SIZE;

    // Write slot metadata: offset, capacity (unused for agg, set 0), size (unused), type_flags
    std.mem.writeInt(u32, state[meta_base..][0..4], slot_offset, .little);
    std.mem.writeInt(u32, state[meta_base + 4 ..][0..4], 0, .little);
    std.mem.writeInt(u32, state[meta_base + 8 ..][0..4], 0, .little);
    state[meta_base + 12] = @intFromEnum(types.SlotType.AGGREGATE);

    // Write initial value=10.0 at offset, count=5 at offset+8
    const val_ptr: *u64 = @ptrCast(@alignCast(&state[slot_offset]));
    const count_ptr: *u64 = @ptrCast(@alignCast(&state[slot_offset + 8]));
    const initial_val: f64 = 10.0;
    val_ptr.* = @bitCast(initial_val);
    count_ptr.* = 5;

    // Save previous state for rollback
    const prev_val_bits: u64 = @bitCast(initial_val);
    const prev_count: u32 = 5;

    // Simulate an aggregate update — overwrite to new values
    const new_val: f64 = 25.0;
    val_ptr.* = @bitCast(new_val);
    count_ptr.* = 8;

    // Verify the overwrite took effect
    const read_val: f64 = @bitCast(val_ptr.*);
    try testing.expectApproxEqAbs(@as(f64, 25.0), read_val, 0.001);
    try testing.expectEqual(@as(u64, 8), count_ptr.*);

    // Rollback using the helper
    const meta = types.getSlotMeta(&state, 0);
    rollbackAggUpdate(&state, meta, prev_count, prev_val_bits);

    // Verify value restored to 10.0 and count to 5
    const restored_val: f64 = @bitCast(val_ptr.*);
    try testing.expectApproxEqAbs(@as(f64, 10.0), restored_val, 0.001);
    try testing.expectEqual(@as(u64, 5), count_ptr.*);
}

test "rollback — COUNT_UPDATE restores previous u64 count" {
    var state: [4096]u8 align(8) = [_]u8{0} ** 4096;
    const meta_base = types.STATE_HEADER_SIZE;
    // COUNT slot: 8 bytes (u64 count only), offset just past metadata
    const slot_offset: u32 = types.STATE_HEADER_SIZE + types.SLOT_META_SIZE;

    // Write slot metadata
    std.mem.writeInt(u32, state[meta_base..][0..4], slot_offset, .little);
    std.mem.writeInt(u32, state[meta_base + 4 ..][0..4], 0, .little);
    std.mem.writeInt(u32, state[meta_base + 8 ..][0..4], 0, .little);
    state[meta_base + 12] = @intFromEnum(types.SlotType.AGGREGATE);
    state[meta_base + 13] = @intFromEnum(types.AggType.COUNT);

    // Write initial count=42 at offset
    const count_ptr: *u64 = @ptrCast(@alignCast(&state[slot_offset]));
    count_ptr.* = 42;

    const prev_count: u32 = 42;

    // Simulate a count update — overwrite to new value
    count_ptr.* = 99;
    try testing.expectEqual(@as(u64, 99), count_ptr.*);

    // Rollback using the helper
    const meta = types.getSlotMeta(&state, 0);
    rollbackCountUpdate(&state, meta, prev_count);

    // Verify count restored to 42
    try testing.expectEqual(@as(u64, 42), count_ptr.*);
}

File: packages/columine/crates/columine-vm/src/undo_log.rs
//! Typed rollback helpers over `FlatTable`.
//!
//! Bitmap and TTL rollback remain in the VM module; this module owns the
//! container rollback operations and the serialized undo-entry contract.
//! `FlatUndoOp`/`FlatUndoEntry` are explicit byte layouts used by exported
//! delta and fork segments. In-memory rings, overflow snapshots, change-flag
//! save/restore, and delta export assembly remain VM-owned.

use crate::bytes;
use crate::hash_table::{ENTRY_NONE, ENTRY_U32, FlatTable};
use crate::meta::SlotMetaView;
use columine_types::types::SlotMetaOffset;

/// Operation byte of a serialized undo entry. Discriminants are wire values
/// inside exported delta/fork segments.
#[repr(u8)]
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum FlatUndoOp {
    /// Rollback: tombstone key, decrement size.
    MapInsert = 1,
    /// Rollback: restore prev value + timestamp.
    MapUpdate = 2,
    /// Rollback: restore key + value + timestamp, increment size.
    MapDelete = 3,
    /// Rollback: tombstone elem, decrement size.
    SetInsert = 4,
    /// Rollback: restore elem, increment size.
    SetDelete = 5,
    /// Rollback: restore prev value (f64/i64) + count (u64) — 16-byte slots.
    AggUpdate = 6,
    /// Rollback: tombstone the derived-fact identity at `prev_value`.
    FactInsertNew = 7,
    /// Restore a derived-fact identity and value at `prev_value`.
    FactInsertUpdate = 8,
    /// Restore a retracted derived-fact identity and value at `prev_value`.
    FactRetract = 9,
    /// Rollback: restore list count to prev_value.
    ListAppendUndo = 10,
    /// Rollback: restore prev count (u64) — 8-byte COUNT slot.
    CountUpdate = 11,
    /// Restore a scalar slot's value and comparison timestamp.
    /// Scalar-update is an extension; the frozen ABI did not journal scalar writes.
    ScalarUpdate = 14,
    /// Restore `pad1` (1..=8) raw bytes at absolute state offset `key` from
    /// `aux`'s little-endian bytes. Used as paired before/after entries for
    /// nested arenas and variable payloads without widening this 24-byte ABI.
    StateBytes = 15,
    /// Roll back one struct-map scalar field or remove a newly created row.
    /// `_pad1` is the field index, `_pad2` carries flags, and `aux` stores the
    /// field bytes in little-endian order.
    StructMapField = 12,
    /// Whole-row rollback for struct-map upserts that clear the row bitset.
    /// `_pad2 = 0x02` marks an absent row; `aux` stores up to eight bitset bytes.
    StructMapRow = 13,
}

/// Field-present flag in a `STRUCT_MAP_FIELD` entry.
pub const SMF_BIT_SET: u8 = 0x01;
/// Row-absent flag shared by struct-map rollback entries.
pub const SMF_ROW_ABSENT: u8 = 0x02;
/// `STRUCT_MAP_FIELD`/`STRUCT_MAP_ROW` flags. Both row-absence flags use
/// `0x02` deliberately, so rollback can share the same removal test.
pub const SMR_ROW_ABSENT: u8 = 0x02;

/// One serialized undo/redo entry. Layout: op@0, slot@1, `_pad1`@2,
/// `_pad2`@3, key@4, prev_value@8, four padding bytes, aux@16; size 24.
/// `_pad1`/`_pad2` carry field index and flags for struct-map entries.
/// Following crate convention the struct is plain Rust; the byte contract
/// lives in [`FlatUndoEntry::write_to`] / [`FlatUndoEntry::read_from`], pinned
/// by layout tests.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct FlatUndoEntry {
    pub op: FlatUndoOp,
    /// Slot index (0xFF = derived fact); ignored by STATE_BYTES.
    pub slot: u8,
    /// STRUCT_MAP_FIELD: field_idx. FACT_*: fact_idx low. STATE_BYTES: length.
    pub pad1: u8,
    /// STRUCT_MAP_*: SMF/SMR flags. FACT_*: fact_idx high byte.
    pub pad2: u8,
    /// Container/fact key, or absolute state offset for STATE_BYTES.
    pub key: u32,
    /// Previous value, physical fact slot, or zero for STATE_BYTES.
    pub prev_value: u32,
    /// Previous timestamp bits, target/restored fact value, field/bitset bytes,
    /// or raw STATE_BYTES payload — per-op documented above.
    pub aux: u64,
}

/// Serialized size of one undo entry.
pub const FLAT_UNDO_ENTRY_SIZE: u32 = 24;

/// One undo/redo pair occupying 48 serialized bytes.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct FlatDeltaEntry {
    pub undo: FlatUndoEntry,
    pub redo: FlatUndoEntry,
}

/// Serialized size of a delta pair.
pub const FLAT_DELTA_ENTRY_SIZE: u32 = 48;

impl FlatUndoEntry {
    /// Serialize with the extern-struct layout (see type doc). The four
    /// padding bytes at 12..16 are written as zero.
    pub fn write_to(&self, out: &mut [u8; FLAT_UNDO_ENTRY_SIZE as usize]) {
        out.fill(0);
        out[0] = self.op as u8;
        out[1] = self.slot;
        out[2] = self.pad1;
        out[3] = self.pad2;
        out[4..8].copy_from_slice(&self.key.to_le_bytes());
        out[8..12].copy_from_slice(&self.prev_value.to_le_bytes());
        out[16..24].copy_from_slice(&self.aux.to_le_bytes());
    }

    /// Deserialize; `None` if the op byte is not a `FlatUndoOp` (a corrupt
    /// or foreign segment — an operational condition for segment consumers,
    /// not a programmer bug).
    pub fn read_from(buf: &[u8; FLAT_UNDO_ENTRY_SIZE as usize]) -> Option<Self> {
        let op = match buf[0] {
            1 => FlatUndoOp::MapInsert,
            2 => FlatUndoOp::MapUpdate,
            3 => FlatUndoOp::MapDelete,
            4 => FlatUndoOp::SetInsert,
            5 => FlatUndoOp::SetDelete,
            6 => FlatUndoOp::AggUpdate,
            7 => FlatUndoOp::FactInsertNew,
            8 => FlatUndoOp::FactInsertUpdate,
            9 => FlatUndoOp::FactRetract,
            10 => FlatUndoOp::ListAppendUndo,
            11 => FlatUndoOp::CountUpdate,
            14 => FlatUndoOp::ScalarUpdate,
            12 => FlatUndoOp::StructMapField,
            13 => FlatUndoOp::StructMapRow,
            15 => FlatUndoOp::StateBytes,
            _ => return None,
        };
        Some(Self {
            op,
            slot: buf[1],
            pad1: buf[2],
            pad2: buf[3],
            key: u32::from_le_bytes(buf[4..8].try_into().expect("4-byte slice")),
            prev_value: u32::from_le_bytes(buf[8..12].try_into().expect("4-byte slice")),
            aux: u64::from_le_bytes(buf[16..24].try_into().expect("8-byte slice")),
        })
    }
}

/// Bind the HASHMAP table used by rollback.
fn bind_map(meta: &SlotMetaView) -> FlatTable {
    FlatTable::bind_external(
        meta.offset,
        meta.capacity,
        meta.meta_base + SlotMetaOffset::SIZE,
        ENTRY_U32,
    )
}

/// Bind the HASHSET table used by rollback.
fn bind_set(meta: &SlotMetaView) -> FlatTable {
    FlatTable::bind_external(
        meta.offset,
        meta.capacity,
        meta.meta_base + SlotMetaOffset::SIZE,
        ENTRY_NONE,
    )
}

/// Offset of a HASHMAP timestamp lane: one u64 (f64 bits) per position.
fn ts_off(meta: &SlotMetaView, pos: u32) -> u32 {
    meta.offset + meta.capacity * 8 + pos * 8
}

// Rollback is logical rather than byte-exact: inserts leave TOMBSTONE keys,
// and dead cells retain stale value/timestamp bytes outside the logical table.
/// Roll back a map insertion by writing a tombstone and decrementing size.
pub fn rollback_map_insert(state: &mut [u8], meta: &SlotMetaView, key: u32) -> bool {
    let tbl = bind_map(meta);
    let Some(pos) = tbl.find(state, key) else {
        return false;
    };
    tbl.set_key_at(state, pos, columine_types::types::TOMBSTONE);
    let size = tbl.size(state);
    tbl.set_size(state, size - 1);
    true
}

/// Restore a map value and optional timestamp from raw u64 bits.
pub fn rollback_map_update(
    state: &mut [u8],
    meta: &SlotMetaView,
    key: u32,
    prev_value: u32,
    prev_ts_bits: u64,
) -> bool {
    let tbl = bind_map(meta);
    let Some(pos) = tbl.find(state, key) else {
        return false;
    };
    tbl.set_entry_u32_at(state, pos, prev_value);
    if meta.has_hashmap_timestamp_storage() {
        bytes::write_u64(state, ts_off(meta, pos), prev_ts_bits);
    }
    true
}

/// Restore a deleted map key, value, and optional timestamp at its probe
/// position, incrementing size. Return false when the key is already present
/// or the table has no insertion position.
pub fn rollback_map_delete(
    state: &mut [u8],
    meta: &SlotMetaView,
    key: u32,
    prev_value: u32,
    prev_ts_bits: u64,
) -> bool {
    let tbl = bind_map(meta);
    let Some(probe) = tbl.find_insert(state, key) else {
        return false;
    };
    if probe.found {
        return false;
    }
    tbl.set_key_at(state, probe.pos, key);
    tbl.set_entry_u32_at(state, probe.pos, prev_value);
    if meta.has_hashmap_timestamp_storage() {
        bytes::write_u64(state, ts_off(meta, probe.pos), prev_ts_bits);
    }
    let size = tbl.size(state);
    tbl.set_size(state, size + 1);
    true
}

/// Restore the previous value bits and full u64 count of a 16-byte aggregate
/// slot (`[value: u64][count: u64]`).
// The count uses prev_value (low) and key (high) lanes so values above u32::MAX
// remain lossless.
pub fn rollback_agg_update(
    state: &mut [u8],
    meta: &SlotMetaView,
    prev_count: u64,
    prev_val_bits: u64,
) {
    bytes::write_u64(state, meta.offset, prev_val_bits);
    bytes::write_u64(state, meta.offset + 8, prev_count);
}

/// Restore the count of an 8-byte COUNT-only slot.
// Full u64 count (prev_value low + key high lanes) — see rollback_agg_update.
pub fn rollback_count_update(state: &mut [u8], meta: &SlotMetaView, prev_count: u64) {
    bytes::write_u64(state, meta.offset, prev_count);
}

/// Restore a scalar slot's value bytes and comparison timestamp.
pub fn rollback_scalar_update(state: &mut [u8], meta: &SlotMetaView, value: u64, ts: f64) {
    bytes::write_u64(state, meta.offset, value);
    bytes::write_f64(state, meta.offset + 8, ts);
}

/// Roll back a set insertion by writing a tombstone and decrementing size.
pub fn rollback_set_insert(state: &mut [u8], meta: &SlotMetaView, elem: u32) -> bool {
    let tbl = bind_set(meta);
    let Some(pos) = tbl.find(state, elem) else {
        return false;
    };
    tbl.set_key_at(state, pos, columine_types::types::TOMBSTONE);
    let size = tbl.size(state);
    tbl.set_size(state, size - 1);
    true
}

/// Restore a deleted set element and increment size.
pub fn rollback_set_delete(state: &mut [u8], meta: &SlotMetaView, elem: u32) -> bool {
    let tbl = bind_set(meta);
    let Some(probe) = tbl.find_insert(state, elem) else {
        return false;
    };
    if probe.found {
        return false;
    }
    tbl.set_key_at(state, probe.pos, elem);
    let size = tbl.size(state);
    tbl.set_size(state, size + 1);
    true
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The serialized layout is the contract: op@0, slot@1, `_pad1`@2,
    /// `_pad2`@3, key@4 LE, prev_value@8 LE, zero padding 12..16, aux@16 LE.
    #[test]
    fn flat_undo_entry_serialized_layout_is_pinned() {
        let entry = FlatUndoEntry {
            op: FlatUndoOp::StructMapField,
            slot: 7,
            pad1: 3,
            pad2: SMF_BIT_SET,
            key: 0x11223344,
            prev_value: 0x55667788,
            aux: 0x99aabbccddeeff00,
        };
        let mut buf = [0xa5u8; 24];
        entry.write_to(&mut buf);
        assert_eq!(buf[0], 12); // StructMapField discriminant
        assert_eq!(buf[1], 7);
        assert_eq!(buf[2], 3);
        assert_eq!(buf[3], 0x01);
        assert_eq!(&buf[4..8], &0x11223344u32.to_le_bytes());
        assert_eq!(&buf[8..12], &0x55667788u32.to_le_bytes());
        assert_eq!(&buf[12..16], &[0, 0, 0, 0]); // alignment padding zeroed
        assert_eq!(&buf[16..24], &0x99aabbccddeeff00u64.to_le_bytes());
        assert_eq!(FlatUndoEntry::read_from(&buf), Some(entry));
        assert_eq!(FLAT_UNDO_ENTRY_SIZE, 24);
        assert_eq!(FLAT_DELTA_ENTRY_SIZE, 48);
    }

    /// Every FlatUndoOp discriminant matches the serialized ABI, including the
    /// deliberate gap between LIST_APPEND_UNDO=10 and COUNT_UPDATE=11.
    #[test]
    fn flat_undo_op_discriminants_match_contract() {
        assert_eq!(FlatUndoOp::MapInsert as u8, 1);
        assert_eq!(FlatUndoOp::MapUpdate as u8, 2);
        assert_eq!(FlatUndoOp::MapDelete as u8, 3);
        assert_eq!(FlatUndoOp::SetInsert as u8, 4);
        assert_eq!(FlatUndoOp::SetDelete as u8, 5);
        assert_eq!(FlatUndoOp::AggUpdate as u8, 6);
        assert_eq!(FlatUndoOp::FactInsertNew as u8, 7);
        assert_eq!(FlatUndoOp::FactInsertUpdate as u8, 8);
        assert_eq!(FlatUndoOp::FactRetract as u8, 9);
        assert_eq!(FlatUndoOp::ListAppendUndo as u8, 10);
        assert_eq!(FlatUndoOp::CountUpdate as u8, 11);
        assert_eq!(FlatUndoOp::StructMapField as u8, 12);
        assert_eq!(FlatUndoOp::StructMapRow as u8, 13);
        assert_eq!(FlatUndoOp::ScalarUpdate as u8, 14); // post-parity extension
        assert_eq!(FlatUndoOp::StateBytes as u8, 15); // post-parity extension
        assert_eq!(FlatUndoEntry::read_from(&[0u8; 24]), None); // 0 invalid
        assert_eq!(
            FlatUndoEntry::read_from(&{
                let mut b = [0u8; 24];
                b[0] = 16;
                b
            }),
            None
        ); // beyond the enum
    }
}
